src/config.rs

```diff
@@ -0,0 +1,181 @@
+    #[serde(rename = "format")]
+    pub format_config: FormatConfig,
+    /// Configuration for the `check` and `lint` commands.
+    #[serde(rename = "check")]
+    pub check_config: CheckConfig,
+    /// Common configuration options for all commands.
+    #[serde(rename = "common")]
+    pub common_config: CommonConfig,
+}
+
+/// Represents shared configuration options for Sprocket commands.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case", deny_unknown_fields)]
+pub struct CommonConfig {
+    /// Display color output.
+    pub color: bool,
+    /// The report mode.
+    pub report_mode: Option<Mode>,
+}
+
+impl Default for CommonConfig {
+    fn default() -> Self {
+        Self {
+            color: true,
+            report_mode: None,
+        }
+    }
+}
+
+/// Represents the configuration for the Sprocket `format` command.
+#[derive(Debug, Clone, Default, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case", deny_unknown_fields)]
+pub struct FormatConfig {
+    /// Use tabs for indentation (default is spaces).
+    pub with_tabs: bool,
+    /// The number of spaces to use for indentation levels (default is 4).
+    pub indentation_size: Option<usize>,
+    /// The maximum line length (default is 90).
+    pub max_line_length: Option<usize>,
+}
+
+/// Represents the configuration for the Sprocket `check` and `lint` commands.
+#[derive(Debug, Clone, Default, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case", deny_unknown_fields)]
+pub struct CheckConfig {
+    /// Rule IDs to except from running.
+    pub except: Vec<String>,
+    /// Causes the command to fail if any warnings are reported.
+    pub deny_warnings: bool,
+    /// Causes the command to fail if any notes are reported.
+    pub deny_notes: bool,
+    /// Run the `shellcheck` program on command sections.
+    pub shellcheck: bool,
+    /// Hide diagnostics with `note` severity.
+    pub hide_notes: bool,
+}
+
+impl Config {
+    /// Validate a configuration
+    pub fn validate(&self) -> Result<()> {
+        // Validate the configuration here
+        Ok(())
+    }
+
+    /// Read a configuration file from the specified path.
+    pub fn read_config(path: &str) -> Result<Self> {
+        let data = std::fs::read(path).context("Failed to open config file")?;
+        let text = String::from_utf8(data).expect("Failed to read config file");
+        let config: Config =
+            toml::from_str(text.as_str()).context("Failed to parse config file")?;
+        Ok(config)
+    }
+
+    /// Write a configuration to the specified path.
+    pub fn write_config(&self, path: &str) -> Result<()> {
+        let data = toml::to_string(self).context("Failed to serialize config")?;
+        std::fs::write(path, data).context("Failed to write config file")
+    }
+
+    /// Merge the current configuration with commandline arguments.
+    pub fn merge_args(&self, args: Commands) -> Commands {
```

Comment:
@peterhuene curious if you know of any way to do this that's a bit cleaner. This works, and with the pattern matching we should be getting compile errors if any of the underlying structs change without being updated here. But it's a pretty ugly function.

Reply:
My only suggestion is that we implement a `fn merge(self, other: Self) -> Self` on each of these args types; I think that would clean things up a bit.

Reply:
Wouldn't that require the configuration options to exactly mirror the command line arguments? Right now, we omit some values from the configuration. We also invert at least one argument in the configuration (I think).

Reply:
Oh perhaps just `fn merge(&mut self, config: T)` where `T` is the relevant config type, then?